packages/signer/signer-btc/src/internal/app-binder/task/UpdatePsbtTask.ts

```
@@ -0,0 +1,297 @@
+      return liftEither(this.getUpdatedPsbt(signedPsbt));
+    }).caseOf({
+      Left: (error) => {
+        return CommandResultFactory({
+          error: new UnknownDeviceExchangeError(error),
+        });
+      },
+      Right: (data) => CommandResultFactory({ data }),
+    });
+  }
+
+  private getSignedPsbt(
+    psbt: InternalPsbt,
+    psbtSignatures: PsbtSignature[],
+  ): Either<Error, InternalPsbt> {
+    return Either.encase(() => {
+      for (const psbtSignature of psbtSignatures) {
+        // Note: Looking at BIP32 derivation does not work in the generic case,
+        // since some inputs might not have a BIP32-derived pubkey.
+        const pubkeys = psbt
+          .getInputKeyDatas(psbtSignature.inputIndex, PsbtIn.BIP32_DERIVATION)
+          .map((keyDatas) => {
+            return keyDatas.map((kDataHex) => {
+              const buf = new ByteArrayBuilder();
+              return buf.addHexaStringToData(kDataHex).build();
+            });
+          });
+        pubkeys.map((pkeys) => {
+          if (pkeys.length != 1) {
+            // No legacy BIP32_DERIVATION, assume we're using taproot.
+            const pubkey = psbt
+              .getInputKeyDatas(
+                psbtSignature.inputIndex,
+                PsbtIn.TAP_BIP32_DERIVATION,
+              )
+              .map((keyDatas) => {
+                return keyDatas.map((kDataHex) => {
+                  const buf = new ByteArrayBuilder();
+                  return buf.addHexaStringToData(kDataHex).build();
+                });
+              });
+            pubkey.map((pKey) => {
+              if (pKey.length === 0) {
+                throw new Error(
+                  `Missing pubkey derivation for input ${psbtSignature.inputIndex}`,
+                );
+              }
+              psbt.setInputValue(
+                psbtSignature.inputIndex,
+                PsbtIn.TAP_KEY_SIG,
+                new Value(psbtSignature.signature),
+              );
```

I think it won't cover all the cases, because when signing a script leaf instead of a key, the signature should be put in `PSBT_IN_TAP_SCRIPT_SIG` with the `pubkey` and `tapLeafHash` in the keydata.
But actually that may not be easy to test since we don't support yet complex wallets, and it may make the psbt finalization more complex... So maybe you could just open a ticket to do it later, or even just make that task fail if `psbtSignature.tapleafHash` is set? In that case it would be a known limitation of `SignTransaction` since we already know `SignPsbt` should be used instead for complex flows?

---

Ok very clear, I guess the task should return an error if `tapleafHash` exists for now and we could handle it in another ticket then.